File: backend/app/api/v1/endpoints/forecasting.py
"""Forecasting endpoints."""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.dependencies import get_db, get_forecasting_service
from app.schemas.common import ok
from app.schemas.forecasting import PredictRequest, TrainRequest
from app.services.forecasting_service import ForecastingService

router = APIRouter(prefix="/forecast", tags=["forecasting"])


@router.post("/train")
def train_forecast(
    payload: TrainRequest,
    db: Session | None = Depends(get_db),
    service: ForecastingService = Depends(get_forecasting_service),
) -> dict:
    result = service.train(payload.model_dump(), db=db)
    return ok(result, message="Model trained on a chronological split. Metrics are not a profit guarantee.")


@router.post("/predict")
def predict_forecast(
    payload: PredictRequest,
    service: ForecastingService = Depends(get_forecasting_service),
) -> dict:
    return ok(service.predict(payload.symbol.upper(), payload.model_name, payload.horizon))


@router.get("/compare/{symbol}")
def compare_forecast(
    symbol: str,
    db: Session | None = Depends(get_db),
    service: ForecastingService = Depends(get_forecasting_service),
) -> dict:
    return ok(service.compare(symbol.upper(), db=db))

File: frontend/src/api/forecasting.ts
import { api, ApiEnvelope } from "@/api/client";
import { ForecastPoint, ModelMetrics } from "@/types/forecasting";

export async function trainModel(payload: Record<string, unknown>) {
  const { data } = await api.post<ApiEnvelope<Record<string, unknown>>>("/forecast/train", payload);
  return data.data;
}

export async function predictModel(symbol: string, model_name: string, horizon = 5) {
  const { data } = await api.post<ApiEnvelope<{ predictions: ForecastPoint[]; metrics: Record<string, number> }>>(
    "/forecast/predict",
    { symbol, model_name, horizon },
  );
  return data.data;
}

export async function compareModels(symbol: string) {
  const { data } = await api.get<ApiEnvelope<{ models: ModelMetrics[]; note: string }>>(`/forecast/compare/${symbol}`);
  return data.data;
}
